src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp

```diff
@@ -332,21 +331,131 @@ void VTKPolyDataWriterInterface::writeElementFields( vtkSmartPointer< vtkCellDat
+    //   For now, we apply a special component-by-component treatment to 4-dimension fields (see below), but we do not output fields with 5
+    // dimensions or more.
+
+    // First, check if the field requires a special treatment. For now, array4d< real64 > (for instance, for phaseCompFraction) are
+    // treated separately to make sure that they can be easily visualized in Paraview.
+    int numArrayDims = 0;
+
     er.forElementSubRegions< SUBREGION >( [&]( auto const & esr )
     {
       WrapperBase const & wrapper = esr.getWrapperBase( field );
-      writeField( wrapper, data, esr.size(), count );
+      numArrayDims = wrapper.numArrayDims();
     } );
-    celldata->AddArray( data );
+
+    // Then, output the field to VTK
+    // Note: fields with 5 dimensions or more are not supported and are not output to file below
+    if( numArrayDims < 4 )
+    {
+      // If the field has three dimensions or less, it can be output in a generic fashion
+      writeElementField1D2D3D< SUBREGION >( celldata, er, field );
+    }
+    else if( numArrayDims == 4 )
+    {
+      writeElementField4D< SUBREGION >( celldata, er, field );
+    }
+    else
+    {
+      GEOSX_ERROR( "Fields with more than 4 dimensions cannot be output to VTK" );
+    }
   }
 }
+
+template< class SUBREGION >
+void VTKPolyDataWriterInterface::writeElementField1D2D3D( vtkSmartPointer< vtkCellData > const celldata,
+                                                          ElementRegionBase const & er,
+                                                          string const & field ) const
+{
+  localIndex count = 0;
+
+  vtkSmartPointer< VTKGEOSXData > data = VTKGEOSXData::New();
+  data->SetNumberOfValues( er.getNumberOfElements< SUBREGION >() );
+  data->SetName( field.c_str() );
+
+  er.forElementSubRegions< SUBREGION >( [&]( auto const & esr )
+  {
+    WrapperBase const & wrapper = esr.getWrapperBase( field );
+    writeField( wrapper, data, esr.size(), count );
+  } );
+  celldata->AddArray( data );
+}
+
+template< class SUBREGION >
+void VTKPolyDataWriterInterface::writeElementField4D( vtkSmartPointer< vtkCellData > const celldata,
+                                                      ElementRegionBase const & er,
+                                                      string const & field ) const
+{
+  // Step 1: we need to figure out the size of the two dimensions of the array that we are going to output component by component
+  localIndex nCompInSecondDim = 0;
+  localIndex nCompInThirdDim = 0;
+  er.forElementSubRegions< SUBREGION >( [&]( auto const & esr )
+  {
+    WrapperBase const & wrapper = esr.getWrapperBase( field );
+    std::type_info const & typeID = wrapper.getTypeId();
+    GEOSX_THROW_IF( typeID!=typeid(array4d< real64 >),
+                    "This function is specific to array4d< real64 >", InputError );
+
+    Wrapper< array4d< real64 > > const & wrapperT = dynamicCast< Wrapper< array4d< real64 > > const & >( wrapper );
```

Comment:
@klevzoff Is there a good way to make this piece of code handle all the possible permutations of the array? Should I test of the possible permutations of `array4d`, or is there a way to make this more generic?

Reply:
I don't actually have any simple ideas at the moment. The way we used to do it was with `arrayTypeLambda()`, but that clearly is not scalable - we'd need to add 24 switch cases just for 4D arrays of one data type (`real64`). Although if we're pretty sure about never going above 4D, we could just bite the bullet and do it once and for all (with the caveat that all code that currently relies on `arrayTypeLambda()` needs to be checked/updated to make sure it can handle 4D).

Short of that, I can only think of the following:
* `reinterpret_cast` the wrapper to type `Wrapper< array4d< real64 > >` (note, not `dynamicCast` since we want to avoid the actual `typeid` check)
* move data to host
* convert it to an `arraySlice4d< real64 const >` by calling `toSliceConst()`
* `reinterpret_cast` the result to `arraySlice4d< real64 const, -1 >` (i.e. forcing `USD=-1`). 

This is technically undefined behavior, but it should work in practice. It relies on the fact that array objects have exact same binary representation regardless of layout - only the computed stride values differ. It slightly pessimizes offset computation (since it has to multiply all strides in absence of single stride dim), but it's largely inconsequential for output. Still, this is brittle and might break in the future - imagine we "optimized" arrays by squashing `m_strides` (i.e. not physically store the unit stride).

@rrsettgast @corbett5 opinions?

Reply:
Thanks @klevzoff. For now I implemented what you suggested with the `reinterpret_cast` and I will backtrack if we find a better solution. The current version is working, although I have not tested on GPU yet, and I don't have a good test for a permuted `array4d` (but I guess your PR will do that soon).

Reply:
I think adding in a constructor for a -1 strided slice from one with a known slice would be a great call. I'll make an issue, I think it should be pretty straightforward but I doubt I'll get around to it in time for this PR.

Reply:
Thanks a lot @corbett5. I ended up using a different approach based on `applyArrayTypeLambda2` to figure out the type of the `array4d`, and now I have something relatively clean that should support all the permutations without requiring the `reinterpret_cast`. So I think there is no longer an immediate need for this constructor in the present PR.

Reply:
Did you just add all the types (or at least all the types we're currently using) to the giant switch yard? If so I'd much rather have the generic slice and get rid of them. I think most of those dispatchers are used for IO anyways.

Reply:
Yes in the current version of the PR, all the permutations of `array4d` are added to the dispatch in `applyArrayTypeLambda2`, which requires some changes in `common/FieldSpecificationOps.hpp`. But if the generic slice is a cleaner option, that is perfectly fine with me.

Reply:
Well the issue is that now when you pass something to `applyArrayTypeLambda2` it has to instantiate all 24 different variants, which really bloats compile time. God forbid we ever do 5D...

Reply:
I'm not at all saying that you should wait, I think this is a good solution. I've just been trying to ditch those type dispatchers for ever, I think they're quite nasty.